src/main/java/ml/shifu/shifu/core/Scorer.java

```diff
@@ -169,85 +177,50 @@ public ScoreObject scoreNsData(MLDataPair inputPair, Map<NSColumn, String> rawNs
-
-                        if(fnlOutputHiddenLayerIndex == 0) {
-                            return finalOutput;
-                        }
-
-                        // append output values in hidden layer
-                        double[] hiddenOutputs = network.getLayerOutput(fnlOutputHiddenLayerIndex);
-                        double[] outputs = new double[finalOutput.getData().length + hiddenOutputs.length];
+                final boolean isOutputFirstHiddenLayer = outputFirstHiddenLayer;
+                    MLData finalOutput = network.compute(networkPair.getInput());
+                    if(!isOutputFirstHiddenLayer) {
+                        modelResults.add(finalOutput);
+                    } else {
+                        // append output values in first hidden layer
+                        double[] firstHiddenOutputs = network.getLayerOutput(1);
+                        double[] outputs = new double[finalOutput.getData().length + firstHiddenOutputs.length];
 
                         System.arraycopy(finalOutput.getData(), 0, outputs, 0, finalOutput.getData().length);
-                        System.arraycopy(hiddenOutputs, 0, outputs, finalOutput.getData().length, hiddenOutputs.length);
-                        return new BasicMLData(outputs);
+                        System.arraycopy(firstHiddenOutputs, 0, outputs, finalOutput.getData().length,
+                                firstHiddenOutputs.length);
+                        modelResults.add(new BasicMLData(outputs));
                     }
-                }.call());
             } else if(model instanceof BasicNetwork) {
                 final BasicNetwork network = (BasicNetwork) model;
 
                 final MLDataPair networkPair = CommonUtils.assembleNsDataPair(binCategoryMap, noVarSelect, modelConfig,
                         columnConfigList, rawNsDataMap, cutoff, alg, null);
-                modelResults.add(new Callable<MLData>() {
-                    @Override
-                    public MLData call() {
-                        return network.compute(networkPair.getInput());
-                    }
-                }.call());
+                modelResults.add(network.compute(networkPair.getInput()));
             } else if(model instanceof SVM) {
                 final SVM svm = (SVM) model;
                 if(svm.getInputCount() != pair.getInput().size()) {
                     log.error("SVM and input size mismatch: SVM Size = " + svm.getInputCount() + "; Input Size = "
                             + pair.getInput().size());
                     continue;
                 }
-                modelResults.add(new Callable<MLData>() {
-                    @Override
-                    public MLData call() {
-                        return svm.compute(pair.getInput());
-                    }
-                }.call());
+                modelResults.add(svm.compute(pair.getInput()));
```

Comment:
good to refine, while in fact I would like to keep Callable which would like to make a configration to switch from single threading to multiple threading version, can we merge into shifu core based code.